### PR TITLE
Fix: memory leaks

### PR DESCRIPTION
**Pull request by gaebel**
- Fixes the memory leaks (#25)
- Makes set function value parameter more verbose
- Clear now returns boolean
- Refactoring iOS
- Documentation

### CODE REVIEW

File: kvault/src/iosMain/kotlin/com/liftric/kvault/KVault.kt [hunk 1]
```diff
@@ -20,40 +18,36 @@ actual open class KVault(
...
      */
-    @Deprecated("Will be removed in a future version, please use the primary constructor. Check your service name before migrating.")
+    @Deprecated(
+        """
+            Will be removed in a future version, 
+            please use the primary constructor. 
+            Check your service name before migrating.
+            """
+        ,
+        level = DeprecationLevel.WARNING
```
Comment (gaebel) [inline kvault/src/iosMain/kotlin/com/liftric/kvault/KVault.kt:32]: Maybe `DeprecationLevel.ERROR`, or too early?

File: iostests/Sources/DebugViewController.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        self.performOperations(on: vault)
+    }
+
+    private func performOperations(on vault: KVault) {
+        vault.set(key: "A", value: true)
+        vault.set(key: "B", value_: 1)
+        vault.set(key: "C", value__: 1)
+        vault.set(key: "D", value___: 1)
+        vault.set(key: "E", value____: 1)
+        vault.set(key: "F", value_____: "T")
```
Comment (benjohnde) [inline iostests/Sources/DebugViewController.swift:42]: That notion is here weird. We could think about a potential refactoring of the API for having a smoother usage in Swift.

File: kvault/src/iosMain/kotlin/com/liftric/kvault/KVault.kt (hunk 1 above)
Comment (benjohnde) [reply kvault/src/iosMain/kotlin/com/liftric/kvault/KVault.kt:32]: I think `WARNING` is just about perfect here!

File: kvault/src/iosMain/kotlin/com/liftric/kvault/KVault.kt [hunk 3]
```diff
@@ -190,141 +158,147 @@ actual open class KVault(
...
-                return true
+    actual fun existsObject(forKey: String): Boolean = memScoped {
+        retain(forKey) { (key) ->
+            makeQuery(
+                kSecClass to kSecClassGenericPassword,
+                kSecAttrAccount to key,
+                kSecReturnData to kCFBooleanFalse
+            ) { query ->
+                SecItemCopyMatching(query, null)
+                    .release(query)
```
Comment (benjohnde) [inline kvault/src/iosMain/kotlin/com/liftric/kvault/KVault.kt:169]: Interesting, that this one now releases the whole memory usage.

**Review by benjohnde — APPROVED** [review]
LGTM @gaebel!

File: kvault/src/iosMain/kotlin/com/liftric/kvault/KVault.kt (hunk 3 above)
Comment (gaebel) [reply kvault/src/iosMain/kotlin/com/liftric/kvault/KVault.kt:169]: The CFBridgingRetain in the CFDictionary does also need to get released.

File: iostests/Sources/DebugViewController.swift (hunk 2 above)
Comment (gaebel) [reply iostests/Sources/DebugViewController.swift:42]: Yeah, I don't know why the Kotlin to Swift conversion ends up like this. The same function with different parameter types should work. We could change it to intValue, stringValue, ....

**Comment by gaebel** [comment]
Will refactor a bit more

**Comment by gaebel** [comment]
@benjohnde This can be merged now. If you want you could look into the single memory leak that occurs using a background queue.

**Pull request merged**